### PR TITLE
Allow option to have debug symbols in a folder for Linux OS

### PR DESCRIPTION
**Pull request by HanesSciarrone**
|Related issue|
|---|
|#12623|

## Description

This PR aims to generate debug symbols inside of `symbols` folder when the option `DISABLE_STRIP_SYMBOLS` is disabled. After the change, some tests were done and the result of the binaries size was:

### SERVER
<details>
<summary>DISABLE_STRIP_SYMBOLS enable</summary>

image

 </details>

<details>
<summary>DISABLE_STRIP_SYMBOLS disable</summary>

image

 </details>

<details>
<summary>Debug symbols</summary>

image

</details>

### AGENT

<details>
<summary>DISABLE_STRIP_SYMBOLS enable</summary>

image

 </details>

<details>
<summary>DISABLE_STRIP_SYMBOLS disable</summary>

image

 </details>

<details>
<summary>Debug symbols</summary>

image

</details>

- Compilation without warnings in every supported platform
  - [x] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [ ] Source installation
- [ ] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

- Decoder/Rule tests
  - [ ] Added unit testing files ".ini"
  - [ ] runtests.py executed without errors

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -86,6 +86,13 @@ DISABLE_SYSC?=no
 DISABLE_CISCAT?=no
 DISABLE_STRIP_SYMBOLS?=no
 
+ifeq (${uname_S}, Linux)
+EXECUTABLE_FILES := $(shell find . -maxdepth 1 -type f ! -name "*.so"  -perm /a=x)
+EXECUTABLE_FILES += $(shell find . -type f -not -path "./external/*" -name "*.so" -perm /a=x)
+EXECUTABLE_FILES:= $(EXECUTABLE_FILES:./%=%)
+endif
```
Comment (jnasselle) [inline src/Makefile:93]: Winagent is compiled also in Linux OS, so it is needed to also filter by target
```suggestion
ifeq (${uname_S}, Linux)
ifneq (${TARGET}, winagent)
EXECUTABLE_FILES := $(shell find . -maxdepth 1 -type f ! -name "*.so"  -perm /a=x)
EXECUTABLE_FILES += $(shell find . -type f -not -path "./external/*" -name "*.so" -perm /a=x)
EXECUTABLE_FILES:= $(EXECUTABLE_FILES:./%=%)
endif
endif
```

File: src/Makefile [hunk 2]
```diff
@@ -745,6 +752,7 @@ ifneq (${uname_S},HP-UX)
 	${MAKE} ${BUILD_CMAKE_PROJECTS}
 endif
 	${MAKE} ${BUILD_SERVER}
+	@${MAKE} strip DISABLE_STRIP_SYMBOLS=${DISABLE_STRIP_SYMBOLS}
```
Comment (jnasselle) [inline src/Makefile:755]: TARGET env should be also populated

File: src/Makefile [hunk 3]
```diff
@@ -754,6 +762,7 @@ ifneq (${uname_S},HP-UX)
...
```
Comment (jnasselle) [inline src/Makefile:765]: ditto

File: src/Makefile [hunk 4]
```diff
@@ -763,6 +772,7 @@ ifneq (${uname_S},HP-UX)
...
```
Comment (jnasselle) [inline src/Makefile:775]: ditto

File: src/Makefile [hunk 5]
```diff
@@ -772,6 +782,7 @@ ifneq (${uname_S},HP-UX)
 	${MAKE} ${BUILD_CMAKE_PROJECTS}
 endif
 	${MAKE} ${BUILD_AGENT}
+	@${MAKE} strip DISABLE_STRIP_SYMBOLS=${DISABLE_STRIP_SYMBOLS}
```
Comment (jnasselle) [inline src/Makefile:785]: Ditto.
Also keep in mind that macos TARGET also will execute this. @RamiroRD eyes here

File: src/Makefile [hunk 6]
```diff
@@ -785,6 +796,17 @@ $(SELINUX_POLICY): $(SELINUX_MODULE)
...
+ifeq (${TARGET}, agent)
+strip: ${BUILD_AGENT} ${BUILD_CMAKE_PROJECTS}
+else
+strip: ${BUILD_SERVER} ${BUILD_CMAKE_PROJECTS}
+endif
+ifeq (${uname_S}, Linux)
+ifeq (,$(filter ${DISABLE_STRIP_SYMBOLS},YES yes y Y 1))
+	@$(foreach var, ${EXECUTABLE_FILES}, objcopy --only-keep-debug $(var) symbols/$(basename $(notdir $(var))).debug; objcopy --strip-unneeded $(var);)
+endif
+endif
```
Comment (jnasselle) [inline src/Makefile:808]: A few things
- We need to mix TARGET and uname_S to determine the specific dependencies for strip target
- Ditto to the content of strip target

**Review by jnasselle — CHANGES_REQUESTED** [review]

File: src/Makefile (hunk 1 above)
Comment (RamiroRD) [reply src/Makefile:93]: But wouldn't the executable files have to be looked up in a different way? i.e. by filtering .dlls and .exes?

**Review by RamiroRD — APPROVED** [review]

File: src/Makefile [hunk 7]
```diff
@@ -785,6 +798,19 @@ $(SELINUX_POLICY): $(SELINUX_MODULE)
 $(SELINUX_MODULE): $(SELINUX_ENFORCEMENT)
 	checkmodule -M -m -o $@ $?
 
+ifeq (${TARGET}, agent)
+strip: ${BUILD_AGENT} ${BUILD_CMAKE_PROJECTS}
+else
```
Comment (jnasselle) [inline src/Makefile:803]: Add ifeq with TARGET to avoid winagent and server overlap

Comment (jnasselle) [reply src/Makefile:803]: Nevermind. ifneq with winagent will be a propper approach due other targets (hybrid, local)

**Pull request merged**